README.md
# Documentation

**shoogar** is a minimalistic framework for **Node.js** that makes building high-performance web applications very simple. Especially made for people who do not like to deal with controllers - but don't take a pass on the benefits of Node.js..


## Why yet another framework?

Meanwhile, there are a number of useful frameworks for Node.js (see **[node modules][]**). Most of them are (H)MVC-based and inspired by popular  frameworks like `Rails`, `Sinatra`, `Grails`, `Django`, `CodeIgniter` or `Kohana`. Great for Back-End Experts, but no specifically framework for Font-End Developers.

**So here comes shoogar**:

* simple http static file server
* simple and robust router mechanism
* simple scaffolding mechanism
* basic template mechanism
* strings and static HTML fragments for views
* single script module 
* .. and no controller!


### What is planned for the next release?

* DOM manipulation with 'jsdom'
* installation via npm

## Modules

Shoogar uses '**[sequence.js][]**', a tiny extension for asynchronous processing of javascript functions.

Copy the shoogar.js file in the same directory as your root script (eg. `server.js`) or a directory below and require it with a relative path:

    var server = require('./lib/shoogar');
  
Or you can copy `shoogar.js` to somewhere in your `require.paths` array. Then you can use a global require like:

    var server = require('shoogar');
    
See the **[node docs][]** for more details.


## Server

The command for starting the server is: 

	server.start();
	
.. or in `server.js`:

	var server = require('./lib/shoogar');
	// All other parts here
	server.start();
	
Now open a terminal window in the same directory as your `server.js` script and type:

	$ node server.js
	
That' all, and you can see the `shoogar-started`-message: 

	$ node server.js
	>> shoogar started .. 
	>> Listening on port 8000
	
Go in your browser to http://localhost:8000 and you should see the 'It works!' message - otherweise you have a problem.

<h3 style="text-align:center;">It works! The shoogar HTTP server is running ..</h3>

Nice, but what's going on?

## Scaffolding

Scaffolding (known from Rails, Grails, etc.) allows you to auto-generate an application including:

1. Controller actions for create/read/update/delete (CRUD) operations
2. The necessary views

The shoogar Scaffolding mechanism by contrast is very simple. Shoogar generates:

1. a route for the root document, normally something like index.html (see 'Routes' below).
2. three HTML fragments - the top and bottom of a rudimentary HTML document and also the 'It works!' message.

The HTML-Top-Fragment:

	<html>
	  <head>
	    <title>shoogar works!</title>
	  </head>
	  <body>

The HTML-Bottom-Fragment:

	  </body>
	</html>

The message:

	<h3 style="text-align:center;padding: 30px">
		It works! The Shoogar HTTP server is running .. 
	</h3>

Note: The Scaffolding mechanism is enabled by default. 



## Views
	
If you want to override the scaffold view, you can create a new HTML-Top-Fragment from a string:

	server.setTemplates({
	  top : 
		'<!DOCTYPE HTML>'
	    '<html>' + 
	    '  <head>' +
	    '    <title>My template</title>' +
	    '    <style>' +
	    '    body { background: #000; color: #FFF; }' +
	    '    </style>' +
	    '  </head>' +
	    '<body>'
	});	
	
Or you can put the HTML in a Template file:

	<!DOCTYPE HTML>
	<html> 
	  <head>
	    <title>My template</title>
	    <style>
	    body { background: #000; color: #FFF; }
	    </style>
	  </head>
	<body>

and include it:

	server.setTemplates({
	  top : 'templates/header.tmpl'
	});
	
If you need another HTML-Bottom-Fragment, you can do it in the same way:

	server.setTemplates({
	  bottom : 'templates/footer.tmpl'
	});

Allowed file extensions for HTML fragments:

* '.inc' (include files)
* '.tpl' or '.tmpl' (template files)
* '.fhtm' or '.fhtml' (HTML fragment file)

## Routes

In shoogar is a route an URL matching pattern paired with a content type in the form:

	URL-PATTERN : CONTENT-TYPE

<table>
    <tr>
        <th>Matching pattern</th>
        <th>Content type</th>
    </tr>
    <tr>
        <td>String that ends with '.inc', '.tpl'. '.tmpl', '.fhtm' or '.fhtml'</td>
        <td>Relative path to an HTML fragment</td>
    </tr>
    <tr>
        <td>String that ends with '.htm' or '.html'</td>
        <td>Relative path to an HTML document</td>
    </tr>
    <tr>
        <td>String that ends without one of the above extensions</td>
        <td>Pur text or an HTML fragment</td>
    </tr>
    <tr>
        <td>Function</td>
        <td>Function for client side action</td>
    </tr>
</table>


The shoogar scaffolding mechanism auto-generates a route for the root URL pattern '/' and the default page content - the 'It works!' message.

To override this route, do something like this:

	server.setRoutes({
	  '/' : '<h1>Welcome!</h1>'	
	});

Working with strings is not convenient. Better put it in a file and define here the path to the fragment file:

	'/' : 'views/welcome.tmpl'

You can also define a static file:

	'/contact' : 'static/contact.htm'

You like bizarre URLs? No problem:

	'/phone/0/129/40/55/333' : 'views/phone.tmpl'

And you can do it 'RESTfully':

	'/hello/*' : function(){
	  document.write(
	    '<h1>Hello ' + (arguments[0] || ' World') + '!</h1>' +
	  );
	}
	// Example: 
	// http://0.0.0.0:8000/hello/Steve 
	// => <h1>Hello Steve!</h1>
	
	
## Port

You can define a specific port: 

	server.setPort(8001);

Note: this is optional. The default port is 8000.


## Error messages

You can show internal error messages (from Node.js) on page: 

	server.messages(true);

Note: this is optional, disabled per default.


[node docs]: http://nodejs.org/api/modules.html
[node modules]: https://github.com/ry/node/wiki/modules
[sequence.js]: http://github.com/up/sequence


CHANGELOG.md
# ChangeLog

## v0.1.2
### Added Response-Header (cache-control, ..) - 2013/01/03

## v0.1.1
### Changed from module 'async' to 'sequence' - 2012/12/31

## v0.1.0
### First public release - 2011/01/13




lib/shoogar.js
/*
shoogar - a minimalistic framework for Node.js
Version: 0.1.1

Copyright (c) 2010 - 2012 Uli Preuss <me@ulipreuss.eu>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
*/

/*global require: false, exports: false, process: false, console: false, setTimeout: false */

(function () {

  "use strict";

  /* Module dependencies */
  var 
    http = require("http"),
    url = require("url"),
    path = require("path"),
    fs = require("fs"),
    sequence = require("./sequence").sequence,
    shoogar
  ;

  shoogar = {
    routes: [],
    routelength: 0,
    services: [],
    port: 8000,
    base: '',
    document: '',
    ext: {
      fragments: [
        "inc",          // Include file
        "tpl", "tmpl",  // Template file
        "fhtm", "fhtml" // HTML Fragment file
      ],
      staticfiles: [
        "htm", "html"
      ]
    },
    html: {
      top: '',
      bottom: ''
    },
    tmpl: {
      top: '',
      bottom: '',
      E403: 'Error 403',
      E404: 'Error 404'
    },
    errormsg: false,
    scaffold: {
      enabled: true,
      top: '<html><head><title>Shoogar works!</title></head><body>',
      bottom: '</body></html>',
      route: [
        '/', 
        '<h3 style="text-align:center;padding: 30px">' + 
        '  It works! The Shoogar HTTP server is running ..' + 
        '</h3>'
      ]
    }
  };

  /* Enable error messages*/
  exports.messages = function messages(val) {
    shoogar.errormsg = (val === true) ? true : false;
  };

  /* Set port */
  exports.setPort = function port(newport) {
    shoogar.port = newport;
  };

  /* Set route */
  exports.setRoute = function setRoute(route) {
    shoogar.routes.push([route[0], route[1]]);
  };

  /* Set routes */
  exports.setRoutes = function setRoutes(routes) {

    shoogar.scaffold.enabled = false;

    var route, newroute, asterix;
    for (route in routes) {
      if (route !== undefined) {
        shoogar.routelength++;
        newroute = route;
        asterix = route.match(/\*$/gi);
        if (asterix) {
          newroute = route.substring(0, route.length - 2);
          shoogar.services.push(newroute);
        }
        shoogar.routes.push([newroute, routes[route]]);
      }
    }
  };

  /* Set template */
  exports.setTemplates = function setTemplates(templates) {

    var dir = (templates.dir) ? templates.dir : '';

    if (templates.top) {
      shoogar.tmpl.top = dir + templates.top;
    }
    if (templates.bottom) {
      shoogar.tmpl.bottom = dir + templates.bottom;
    }
    if (templates.E403) {
      shoogar.tmpl.E403 = dir + templates.E403;
    }
    if (templates.E404) {
      shoogar.tmpl.E404 = dir + templates.E404;
    }

  };

  exports.start = function start() {

    var filename, content, router, create, get, set, load, ext, server, isInArray;

    if (shoogar.routelength === 0) {
      // Add route for scaffolding
      shoogar.routes.push(shoogar.scaffold.route);
    }

    /* Init router */
    router = function (response, uri) {

      var isRegisteredRoute = isInArray(uri, shoogar.routes, true),
        service = "/" + uri.split('\/')[1],
        parameter = uri.split('\/')[2] || '',
        isRegisteredService = isInArray(service, shoogar.services);

      if (isRegisteredService !== false) {
        isRegisteredRoute = isInArray(service, shoogar.routes, true);
        if (typeof shoogar.routes[isRegisteredRoute][1] === 'function') {
          create(
            response, 
            '<script type="text/javascript">\n' + 
            '(' + shoogar.routes[isRegisteredRoute][1] + '("' + parameter + '"));\n' + 
            '</script>'
          );
        }
      } else if (isRegisteredRoute !== false) {
        create(response, shoogar.routes[isRegisteredRoute][1]);
      } else {

        filename = path.join(process.cwd(), uri);

        fs.exists(filename, function (exists) {

          if (!exists) {
            content = shoogar.tmpl.E404;
            create(response, content);
            return;
          }

          var stats = fs.statSync(filename);
          if (stats.isDirectory()) {
            content = shoogar.tmpl.E403;
            create(response, content);
            return;
          } else if (stats.isFile()) {
            load(response, filename);
            return;
          }

        });
      }

    };

    /* Get extension */
    ext = function (str) {
      var index = str.lastIndexOf(".") + 1;
      return str.substring(index);
    };

    /* Create document */
    create = function (response, content) {

      if (typeof content !== 'function') {
        if (isInArray(ext(content), shoogar.ext.staticfiles)) {
          load(response, content);
          return;
        }
      }

      shoogar.document = '';

      sequence(
        [
          function (callback) {

            // Get HTML top part
            if (shoogar.html.top === '') {
              if (shoogar.tmpl.top !== '') {
                if (isInArray(ext(shoogar.tmpl.top), shoogar.ext.fragments)) {
                  get('top', callback);
                } else {
                  shoogar.html.top = shoogar.tmpl.top;
                  callback();
                }
              } else {
                // Scaffolding view
                if (shoogar.scaffold.enabled) {
                  shoogar.html.top = shoogar.scaffold.top;
                  callback();
                } else if (isInArray(ext(shoogar.tmpl.top), shoogar.ext.fragments)) {
                  get('top', callback);
                }
              }
            } else {
              callback();
            }
          },
          function (callback) {

            // Set HTML Top
            shoogar.document += shoogar.html.top;
            callback();

          },
          function (callback) {

            // Set HTML Content
            if (isInArray(ext(content), shoogar.ext.fragments)) {
              // .. from template file
              set(content, callback);
            } else {
              // .. from string
              shoogar.document += '\n\n' + content + '\n\n';
              callback();
            }
          },
          function (callback) {
            // Get HTML bottom part
            if (shoogar.html.bottom === '') {
              if (shoogar.tmpl.bottom !== '') {
                if (isInArray(ext(shoogar.tmpl.bottom), shoogar.ext.fragments)) {
                  get('bottom', callback);
                } else {
                  shoogar.html.bottom = shoogar.tmpl.bottom;
                  callback();
                }
              } else {
                // Scaffolding view
                if (shoogar.scaffold.enabled) {
                  shoogar.html.bottom = shoogar.scaffold.bottom;
                  callback();
                } else if (isInArray(ext(shoogar.tmpl.bottom), shoogar.ext.fragments)) {
                  get('bottom', callback);
                }
              }
            } else {
              callback();
            }
          },
          function (callback) {
            // Set HTML Bottom
            shoogar.document += shoogar.html.bottom;
            callback();
          }
        ],
        function (callback) {
          // Output
          response.writeHead(200);
          response.write(shoogar.document);
          response.end();
        }
      );

    };

    /* Get content from template */
    get = function (part, callback) {
      var template = fs.createReadStream(shoogar.tmpl[part]);
      template.setEncoding('utf-8');
      
      template.on('data', function (data) {
        data = data.replace(/<\/title\>/, '</title>\n\t<base href="' + shoogar.base + '" />');
        shoogar.html[part] = data;
        return callback();
      });
      template.on('error', function (err) {
        shoogar.html[part] = err.message;
        return callback();
      });
    };

    /* Set content from template */
    set = function (tmpl, callback) {
      var template = fs.createReadStream(tmpl);
      template.setEncoding('utf-8');
      template.on('data', function (data) {
        data = data.replace(/<\/title\>/, '</title>\n\t<base href="' + shoogar.base + '" />');
        shoogar.document += data;
        return callback();
      });
      template.on('error', function (err) {
        shoogar.document += err.message;
        return callback();
      });
    };

    /* Load static file */
    load = function (response, filename) {
      fs.readFile(filename, function (err, file) {

        if (err) {
          if (shoogar.errormsg) {
            content = err.message;
          } else {
            content = shoogar.html.E404;
          }
          create(response, content);
          return;
        }

        response.writeHead(200, { 
          'content-length': file.length,
          'connection': 'keep-alive',
          'cache-control': 'public, max-age=31536000'
        });
        response.write(file);
        response.end();

      });
    };

    /* Array Helper */
    isInArray = function (elem, arr, key) {
      var i;
      for (i = 0; i < arr.length; i++) {
        if (key && arr[i][0] === elem) {
          return i;
        } else if (arr[i] === elem) {
          return true;
        }
      }
      return false;
    };

    /* Start server */
    server = http.createServer();
    server.on("request", function (request, response) {
      var uri = url.parse(request.url).pathname;
      if (shoogar.base === '') {
        shoogar.base = "http://" + request.headers.host;
      }
      router(response, uri);

    }).listen(process.env.PORT || shoogar.port, function () {
      console.log('>> Shoogar started .. ');
      setTimeout(function () {
        console.log('>> Listening on port ' + shoogar.port);
      },
      500);
    });

  };

}());


lib/sequence.js
/*global module: true, define: true, exports: true */

/**
 * Asynchronous processing of functions
 * https://github.com/up/sequence
 * copyright (c) 2012 - 2013 Uli Preuss
 * MIT License
*/

(function () {

  function sequence (tasks, callback) {

    var 
      num = 0, 
      fns = [],
      alen = arguments.length,
      hasTasks
    ;

    hasTasks = (function () {
      if (tasks.constructor.toString().indexOf('Array') === -1) {
        return false;
      }
      else {
        return true; 
      }
    }());
    
    function next() {
      function task(fn) {
        fns[num] = fn;
        if (num === tasks.length && callback) {
          callback.apply(fns, fns);
        } else {
          next();
        }
      }
      if(num < tasks.length) {
        tasks[num++](task);
      }
    }
        
    if(alen === 0) {
      return; //no tasks, no callback
    } else if (alen === 1 && !hasTasks) {
      tasks(); // no tasks, only callback
    } else if (alen === 2 && hasTasks && tasks.length === 0) {
      callback(); // empty task array, only callback
    } else {
      next();  
    }
    
  }

  if (typeof exports !== 'undefined') { 
    exports['sequence'] = sequence;
  } else if (typeof window !== 'undefined') { 
    window['sequence'] = sequence;
  }

}());


server.js

var server = require('./lib/shoogar');

server.setPort(8088);                  // default: 8000

//server.messages(true);               // default: false

/*
server.setTemplates({
  //top : 'application/templates/header.tmpl'
  top : 
    '<html>' + 
    '  <head>' +
    '    <title>My template</title>' +
    '    <style>' +
    '    body { background: #000; color: #FFF; }' +
    '    </style>' +
    '  </head>' +
    '<body>'
});
*/

server.setTemplates({
  dir         : 'application/templates/',   // optional: default is ''
  top         : 'header.tmpl',              // required
  bottom      : 'footer.tmpl',              // required
  E403        : '403.tmpl',                 // optional: default is 'Error 403' (string)
  E404        : '404.tmpl'                  // optional: default is 'Error 404' (string)
});


var awesome = '' +
'<h1>Node.js is totally awesome!</h1>' +
'<h4>Load content from string</h4>' +
'<h5>Route:</h5>' +
'<pre>"/awesome" : "&lt;h1>Node.js is totally awesome!&lt;/h1>"</pre>';

/*
server.setRoutes({
  "/" : "index.htm"	
});
*/

server.setRoutes({
  '/'                      : 'application/views/index.inc', 
  '/awesome'               : awesome,
  '/template'              : 'application/views/template.inc',
  '/changelog'             : 'application/views/changelog.inc',
  '/static'                : 'application/views/static/static.htm',
  '/docs'                  : 'application/views/docs.inc',
  '/phone/0/129/40/55/333' : 'application/views/phone.inc',
  '/hello/*'               : function(){

    var hello = '' +
    '<h4>RESTful URL example</h4>' +
    '<p>Add your name to the Url (eg. /hello/Mike) and press "enter".</p>' +
    '<h5>Route:</h5>' +
    '<pre>"/hello/*" : function(){\n' +
    '  document.write(\n' +
    '    "&lt;h1>Hello " + (arguments[0] || " World") + "!"&lt;/h1>"\n' +
    '  );\n' +
    '}</pre>';

    document.write(
      '<h1>Hello ' + (arguments[0] || ' World') + '!</h1>' +
      hello
    );
  }
});


server.start();

